File: python/113. Path Sum II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(root, sum, [], res)
        return res
    def dfs(self, root, sum, path, res):
        if not root:
            return
        if sum == root.val and not root.left and not root.right:
            path += [root.val]
            res.append(path)
        if root.left:
            self.dfs(root.left, sum-root.val, path + [root.val], res)
        if root.right:
            self.dfs(root.right, sum-root.val, path + [root.val], res)

File: cpp/101-150/113. Path Sum II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<vector<int>> pathSum(TreeNode* root, int sum) {
		vector<vector<int>> res;
		vector<int> path;
		helper(root, sum, path, res);
		return res;
	}
	void helper(TreeNode *root, int sum, vector<int> path, vector<vector<int>> &res) {
		if(!root)
			return;
		if(root->val == sum && !root->right && !root->left) {
			path.push_back(root->val);
			res.push_back(path);
		}
		sum -= root->val;
		path.push_back(root->val);
		helper(root->left, sum, path, res);
		helper(root->right, sum, path, res);
		path.pop()
		return;
	}
};